### ISSUE TITLE
android native-lib.so crash after upx compress

### ISSUE THREAD

**Issue by dingwenjie** [upx/upx#318]
## What's the problem (or question)?
1. compress native-lib.so use upx 3.9.5 with --android-shlib  parameter, compress success. then use the so file which has compressed, crash on android phone
2. use the develop branch source code, compress native-lib.so,  throwCantPack.

## What should have happened?
1. for problem 1,  should not crash
2. for problem 2, should compress success

## Do you have an idea for a solution?
1. I have no idea with problem 1.
2. for problem 2. 

- in p_unix.cpp

change 
`
#define __S_IEXEC 0100  /* Execute by owner.  */ 
`
to
`
#define __S_IEXEC 0x0100  /* Execute by owner.  */
`

- in p_lx_elf.cpp

add other r_type value
`
if (Elf32_Ehdr::EM_ARM == e_machine
                                    &&  R_ARM_RELATIVE == r_type) {
                                        user_init_va = get_te32(&file_image[user_init_off]);
                                    }
                                    else {
                                        char msg[50]; snprintf(msg, sizeof(msg),
                                            "bad relocation %#x DT_INIT_ARRAY[0]",
                                            r_info);
                                        throwCantPack(msg);
                                    }
                                    break;
 `
in my case r_type is number 2
`
if (Elf32_Ehdr::EM_ARM == e_machine  &&  (R_ARM_RELATIVE == r_type || xxxx == r_type))
`

## How can we reproduce the issue?

1. create new native project use android studio
2. modify the native-lib.c :
`
#include <jni.h>
#include <string>
#include <android/log.h>

void testBeforeMain101() __attribute__((constructor(101)));

extern "C" JNIEXPORT jint JNICALL
JNI_OnLoad(JavaVM *vm, void *reserved) {
    __android_log_print(ANDROID_LOG_DEBUG, "demo", "eyas JNI_OnLoad\n");
    return JNI_VERSION_1_4;
}

extern "C" JNIEXPORT jstring JNICALL
Java_com_example_eyas_MainActivity_stringFromJNI(
        JNIEnv *env,
        jobject /* this */) {
    std::string hello = "Hello from eyas";
    __android_log_print(ANDROID_LOG_DEBUG, "demo", "eyas Java_com_example_reforceapk_MainActivity_stringFromJNI\n");
    return env->NewStringUTF(hello.c_str());
}

void testBeforeMain101() {
    __android_log_print(ANDROID_LOG_DEBUG, "demo", "testBeforeMain 101\n");
}
`
3. build apk, compress native-lib.so, resign apk 
4. install & run the apk

## Please tell us details about your environment.

* UPX 3.9.5 and upx dev branch
* win 10:
* intel:
* armeabi-v7a:
* arm

**Comment by dingwenjie** [comment]
**dev branch** 

android app crash with arm64-v8a so compressed use upx.
in file p_lx_elf.cpp, canPack function, i add this code: 
`
} else if (R_AARCH64_ABS64 == r_type) {
    user_init_va = get_te64(&file_image[user_init_off]);
`
and change the 
`get_te64(&shdr->sh_type)` to `get_te32(&shdr->sh_type)`, 

it compress so success, but crashed while running app with the message:

> I/demo: before loadlibrary nativetest:
> A/libc: Fatal signal 4 (SIGILL), code 1, fault addr 0x7b8afa5000 in tid 8515 (om.example.eyas)
> Disconnected from the target VM, address: 'localhost:8606', transport: 'socket'

so file: [https://pan.baidu.com/s/1zSNLSqlPmcVK7aj9p10uEg](url) code: vukj

i test armeabi-v7a, it is ok. @jreiser

**Comment by jreiser** [comment]
Thank you for the fix `get_te32(&shdr->sh_type)`.  The `R_AARCH64_ABS64` change was made to the `devel` branch.  No progress as yet on the runtime crash.

**Comment by dingwenjie** [comment]
Thank you very much for your reply.

In the past few days, I have tried to solve the crash problem after arm64 hardening. But I don't know about ELF files and upx projects (in fact, I don't know much about them now), so it took me a long time trying to figure out the logic of upx projects, but it hasn't been solved so far.

i found an error:
```
PackLinuxElf64 :: pack1 method line 3683:
                  else {// SHF_ALLOC: in PT_LOAD; but move sh_addr and sh_offset
                        upx_uint64_t pos = xct_off + e_shnum * sizeof (Elf64_Shdr);
                        set_te64 (& shdr-> sh_addr, pos);
                        set_te64 (& shdr-> sh_offset, pos);
                    }
```
  In my test project, I have two note segments.
```
  NOTE 0x0000000000000200 0x0000000000000200 0x0000000000000200
                 0x0000000000000024 0x0000000000000024 R 4
  NOTE 0x0000000000032804 0x0000000000032804 0x0000000000032804
                 0x0000000000000098 0x0000000000000098 R
```
  The "else" code block will change the offset of the section corresponding to my first note segment.
  So I added judgment like this:
```
if (xct_off <= sh_offset) {
upx_uint64_t pos = xct_off + e_shnum * sizeof (Elf64_Shdr);
set_te64 (& shdr-> sh_addr, pos);
set_te64 (& shdr-> sh_offset, pos);
}
```
  I don't know if the logic should be like this, but at least this case of my test project can handle it.

And I have some questions need your help: 

- Question 1: 

In p_lx_elf.cpp file, PackLinuxElf64 :: pack1 method line 3625 and PackLinuxElf32 :: pack1 method line 3266. Do I need to process R_ARM_ABS32 and R_AARCH64_ABS64? In other words, what types need to be dealt with here?

- Question 2:

In the p_lx_elf.cpp file, PackLinuxElf64 :: canPack method line 2529,
why
`user_init_off = (char const *) & dynp-> d_val-(char const *) & file_image [0];`

- Question 3:

Why
```
jni_onload_va = get_te32 (& jni_onload_sym-> st_value);
jni_onload_va = 0;
```
After jni_onload_va gets the value, it is immediately reset to 0. ELF32 and ELF64 are set like this

- Question 4:

The PackLinuxElf32 :: ARM_updateLoader method uses elfout.phdr [0], but when I debug, I find that elfout was uninitialized at the time. All are 0xcdcdcdcd;
The PackLinuxElf64 :: updateLoader method, if it is shlib, returns directly.

- Question 5:

Is it OK for arm64 to set asl_delta to 0x10000?

- Question 6:

In the p_lx_elf.cpp file, PackLinuxElf64 :: pack1 method line 3712:
`sz_elf_hdrs = fpad8 (fo);`

But in PackLinuxElf :: pack3 method line 326:
`unsigned const t = (4 & len) ^ ((!! xct_off) << 2);`

PackLinuxElf :: pack3 method line 363:
`return fpad4 (fo); // MATCH03`

PackLinuxElf64 :: pack3 method line 511:
`flen = fpad4 (fo);`

PackLinuxElf64 :: pack2 method line 4026:
`sz_pack2a = fpad4 (fo); // MATCH01`

Why is the first one to use fpad8 ,the others use fpad4?

### In summary:
In addition to the above problems, ELF64's processing is almost the same as ELF32. Why can't 64 run? (I tried to set asl_delta to 0x10000, and changed all 4 alignments to 8 alignments in question 6. It doesn't seem to have any effect.) It is impossible that there is something wrong with the ELF64 loader?

This is my test project: [eyastest](https://github.com/dingwenjie/eyastest)

look forward to hearing from you.

**Comment by jreiser** [comment]
* Question 0: Thank you for the suggested guard `if (xct_off <= sh_offset)'.  I wonder why the SHT_NOTE are not combined into one Shdr.
* Question 1: It is unknown which relocation types must be handled.  I added code to detect unhandled `r_type`.
* Question 2: In this case no DT_INIT* has been seen at all.  We are taking an unused entry from the `preload` space, and converting it to a DT_INIT, so `user_init_off` must be set.
*Question 3: What to do with `jni_onload_va` is not understood; an example is needed.  Meanwhile, remember the places where it should be detected.
* Question 4: Please upload an example which fails.
* Question 5: Setting `asl_delta` to 0x10000 probably can work, but writing the file may need some adjusting in order to have the correct physical offsets.
* Question 6: `fpad8` should be used when one or more following data items are known to have `(8==sizeof)`, in case the hardware requires 8-byte alignment.  However the upx runtime stub only uses 4-byte data items; upx regards files that are larger than 4GiB as grotesque.

Re: eyastest:  Thank you for the link to the project.
So far I have been frustrated by Android development environments.  Android Studio does not even install on my recent Fedora Linux x86_64 system, for instance.  On Ubuntu the download package of Android Studio is missing required files, and finding via searching the next is very tedious.  When the pieces finally are found, then the documentation of how to build a shared library, then overwrite it with a compressed version, is nowhere to be seen.  Also, I am amazed that "crash" reports do not include the values of all registers, a dump of the memory region surrounding the program counter, and other information that has been expected for decades.  How do _you_ construct Android software?

**Comment by dingwenjie** [comment]
Sorry I have no experience with Fedora or Ubuntu, my working environment is always win10 or Mac OS, AndroidStudio is very easy to use on these two systems, as long as the installation is complete, you can use it. It also automatically downloads dependencies after importing the project.

**Comment by dingwenjie** [comment]
I have added screenshots of the environment when question 4 occurred at the link address:
[issue318_question_4_screenshot](https://github.com/dingwenjie/eyastest/blob/master/other_res/upx_issue318_question_4.png)

**Comment by dingwenjie** [comment]
sorry, i seem to have forgotten to say a place, line 778 of p_lx_elf.cpp seems to be get_te32

**Comment by dingwenjie** [comment]
Sorry i just noticed this problem
"the documentation of how to build a shared library, then overwrite it with a compressed version, is nowhere to be seen."
-------------------------
 Switching between the red code block and the blue code block can switch the packaging of the original so and the upx compressed so. The compressed so is placed in the corresponding architecture of the jniLibs directory, and the red code block is enabled to pack the compressed so apk. If you pack the apk with the original so, delete the corresponding so in jniLibs, and enable the blue code block. I use it often in testing, a lazy way

**Comment by dingwenjie** [comment]
hi, Is there any progress about the crash? @jreiser
